recipes/full_finetune_distributed.py

```diff
@@ -633,11 +744,17 @@ def save_checkpoint(
 
         intermediate_checkpoint = epoch + 1 < self.total_epochs
 
+        # If async checkpointing is enabled, intermediate checkpoints will
+        # be saved asynchronously.
+        if intermediate_checkpoint and self._enable_async_checkpointing:
```

Comment:
Is there ever a case where we wouldn't want async checkpointing for intermediate? Is the flag really needed? Also, it should be possible to merge the async save and full save logic into the same function to not proliferate state dict code everywhere.

Reply:
@pbontrager This flag was added temporarily to make it an opt-in feature which allows us to get more validations before we make it a default. Also, in case of any bugs, devs can turn it off and fallback to the prior state. Currently async save and full save happen via different Checkpointers having different formats but we can probably create a helper method to move this logic out of the recipe code. I will refactor this a bit more.